### DISCUSSION TITLE
Forwarding another port for longpolling

### DISCUSSION THREAD

**Discussion by Cl0ut1eR** [nginx-proxy/nginx-proxy#2013] [Q&A] [answered]
I am using odoo with docker. everything was working fine with the automatic config. 

I want to use the port 8072 of the container to enable longpolling on my odoo environment, 
but i cant seem to find the way to add 
```
location /longpolling {
    proxy_pass http://sub.domain.com:8072;
}
```
to the config file.

im asuming its by creating the file sub.domain.com in the `/etc/nginx/vhost.d` dirrectory but its is not working.

here is what the file i created looks like.
```
## Start of configuration add by letsencrypt container
location ^~ /.well-known/acme-challenge/ {
    auth_basic off;
    auth_request off;
    allow all;
    root /usr/share/nginx/html;
    try_files $uri =404;
    break;
}
## End of configuration add by letsencrypt container
location /longpolling {
	proxy_pass http://sub.domain.com:8072;
}
```
UPDATE 
found a part of the problem, the upstream sub.domain.com already contains the port 8069.
but by hard coding the ip of the container, longpolling is working.
```
location /longpolling {
     proxy_pass http://172.18.0.X:8072;
}
```

**Comment by wguerram** [comment]
Hello, would you mind sharing your ngix config file and the mapping you did in the docker volume? Did you find a way to not hardcode the container IP?

**Reply by wguerram** [reply]
I created this file mydomain.com_location in folder config

```
location /longpolling {
     proxy_pass http://10.0.2.72:8072;
}
```

and added this volume

```
- ./config:/etc/nginx/vhost.d
```

I keep getting this error in the logs

```
2022-08-11 13:04:59,963 14 INFO demo werkzeug: db - - [11/Aug/2022 13:04:59] "POST /longpolling/poll HTTP/1.1" 200 - 1 0.001 0.006
2022-08-11 13:05:04,325 15 INFO demo werkzeug: db - - [11/Aug/2022 13:05:04] "POST /longpolling/im_status HTTP/1.1" 200 - 4 0.010 0.015
2022-08-11 13:05:11,069 15 ERROR demo odoo.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 687, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 359, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 348, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 916, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 535, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/bus/controllers/main.py", line 28, in poll
    raise Exception("bus.Bus unavailable")
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 643, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 301, in _handle_exception
    raise exception.with_traceback(None) from new_cause
Exception: bus.Bus unavailable
```

**Comment by wguerram** [comment]
I made it work, I was putting the nginx volume in the domain container itself instead of the nginx-proxy container. The only problem I have is the hardcoded IP.

**Comment by Cl0ut1eR** [comment, marked as answer]
Sorry for the long reply, i made it work by changing the base configuration file. `/app/nginx.tmpl`
i modified the function to create upstreams, to accept a port from the code and called it with a hard codded port. 
in my case this file will never be changed so i does not really matter that it is hard coded. Im sure you could make something really nice by using environment variables, like `SecondPort1 = 123` and `SecondPortName = Name` (its not real code) or something like this.
here is one of my call to add the second port. (the function was changed so it wont work by default for you) 
`{{ template "upstream" (dict "Upstream" $longpolling "Containers" $containers "Networks" $CurrentContainer.Networks "Debug" $debug_all "ForcePort" "8072") }}`
i added the file that i used in the project. 
Check the 
upstream definition starting at line 88, 
upstream call at line 270,
location call at line 395.

the code i added is not really a model to follow, i didint know what i was doing, it was trial and error to understand what did what but it works, it adds a config to every containers for longpooling to port 8072.  Like i said im sure you could tweak this idea to use the environment variables and not add the port blindly to every containers.

[nginx.tmpl.txt](https://github.com/nginx-proxy/nginx-proxy/files/9349255/nginx.tmpl.txt)

**Comment by k0dard** [comment]
I am also looking for a way not to hardcode container IP in virtualhost_location file. Is it possible to refference it inside _location file without modifying nginx.tmpl?

**Reply by Cl0ut1eR** [reply]
I'm no expert, but from what I learned it would be simple to juste change the configuration file and add a environment variable to the container. From there you can do whatever you need without hard coding anything.

If you really want/need to not change the base file, maybe someone else would be of assistance.

**Reply by k0dard** [reply]
Thanks for your fast reply!
Currently I mount host volume to nginx-proxy container, like --volume /etc/nginx/vhost.d:/etc/nginx/vhost.d. Inside this directory is my virtual host config file with some contents:
```
location /myloc/ {
  proxy_pass 172.17.0.4:3003;
}
```
This seems to work, but I would prefer to have some virtual_host_ip variable instead of hardcoded ip of the app container.
I don't see a way to get virtual host (container) IP address from inside nginx-proxy container. Any idea?

**Reply by Cl0ut1eR** [reply]
Unless your recreate the file with code to change the port I don't see how you could do it. But I finished my project with that so I didn't continue to learn how it was working.

**Reply by JonasBVS** [reply]
i made it work by replacing the ip with the container name:

composefile:
```
services:
  container_name:
    [...]
```

/etc/nginx/vhost.d/mydomain.com_location:
```
location /longpolling {
     proxy_pass http://container_name:8072;
}
```
